static/skin/i18n.js

```diff
@@ -73,11 +73,30 @@ const DEFAULT_UI_LANGUAGE = 'en';
 
 Translations.load(DEFAULT_UI_LANGUAGE, /*asDefault=*/true);
 
+// Below function selects the most suitable UI language from the list
+// of preferred languages in browser preferences and available translations.
+// Since, unlike Accept-Language header, navigator.languages doesn't contain
+// qvalues, they are assumed to decrease exponentially.
+function getDefaultUserLanguage() {
+  const mostSuitableLang = { code: DEFAULT_UI_LANGUAGE, score: 0 }
+  for (const lang of uiLanguages ) {
+    const rank = navigator.languages.indexOf(lang.iso_code);
+    if ( rank >= 0 ) {
+      const pseudoQvalue = Math.pow(0.9, rank);
+      const score = pseudoQvalue * lang.translation_count;
+      if ( score > mostSuitableLang.score ) {
+        mostSuitableLang.code = lang.iso_code;
+        mostSuitableLang.score = score;
+      }
+    }
+  }
+  return mostSuitableLang.code;
+}
```

Comment:
I like the idea of using the translated lines counts to select the best language.

But why with formula ? Do you have a known use case where the formula works (or not) ?

We can be sure that we will have issue opened about "My setting are not respected and kiwix use another language". I prefer to make the first move and prepare an answer 🙂

Reply:
I agree.

Reply:
This reproduces the algorithm used in the backend. The only difference is that qvalues are made up.

Consider the following language list: `ab`, `bc`, `cd`, `de`, and suppose that translation of `kiwix-serve` UI into those languages is at 30%, 90%, 50% and 100%, respectively. If we assume no qvalues (or, equivalently, qvalues equal to 1 for all languages) this actually means that the order of languages never matters and we have to select the language with the highest level of translation (in this case `de` which is at the last position). Therefore, in order to account for the order in which the languages are listed we must assume a decreasing sequence of qvalues which in my implementation is 1, 0.9, 0.81, ... 0.9**k, .... In this example `bc` is the winner with a score of 0.81 vs 0.729 of `de`.

Of course the usage of `0.9` as the common ratio of the geometric progression used for  qvalues is arbitrary and `0.8` or `0.7` may be more appropriate. Note however that exponential functions with a base less than 1 decrease quite rapidly so the utility for our purpose of other values with a single digit after decimal point (0.6, 0.5, etc) drops significantly.

Yet nothing forces us to use a formula. I can happily replace it with any hardcoded sequence like:

```python
assumedQvalues = [1, 0.98, 0.76, 0.54, 0.32, ...]
```

Reply:
FYI, in Firefox 121, the qvalue sequence implicitly generated by the browser depends on the number of languages in Webpage Language Settings. It is
- `[1, 0.5]` for 2 entries
- `[1, 0.7, 0.3]` for 3 entries
- `[1, 0.8, 0.5, 0.3]` for 4 entries

Reply:
Their algorithm seems to be
```python
[round(1-i/n, 1) for i in range(n)]
```

under the assumption that values at the middle are rounded upward (i.e. `0.25` is rounded up to `0.3` rather than down to `0.2`)

Reply:
Seems good to me.